go/vt/vtgate/plugin_mysql_server.go

```diff
@@ -225,6 +243,30 @@ func init() {
 			mysqlUnixListener.Close()
 			mysqlUnixListener = nil
 		}
+		busyLock.Lock()
+		defer busyLock.Unlock()
+		if len(busy) == 0 {
+			log.Info("No mysql connections to drain.")
+			return
+		}
+		log.Infof("Draining %v mysql connection(s)...", len(busy))
+		for {
+			var notFinished bool
+			for c, connIsBusy := range busy {
+				if !connIsBusy {
```

Comment:
It seems like there's a race when you close the connection out from under it, since another thread could be about to start a new transaction.

Reply:
The race is harmless: If a transaction (or request) is started after connection close, it will get added back to the map, and the outer loop will just re-close it.

But I like the other suggestion of just tracking count, which will completely eliminate this race.